### PR TITLE
90% TENT-1278 Support for delta tokens and HEAD requests

### PR DESCRIPTION
**Pull request by rgubby**
We currently don't support delta_tokens in GET requests to feeds - this rectifies that.

Also, this PR introduces the ability to make a HEAD request to see if there are any new posts, with or without a delta_token.

#### Todo
* [x] Bit of repetition with getting params + constructing a URL - parametize it
* [x] Update documentation on how to use the new features
* [x] Unit tests

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -35,15 +35,69 @@ var BabelClient = function (config) {
...
+        params = {};
+    }
+
+    // Build up a query string
+    var queryString = '',
+        queryStringParams = [];
+
+    if(!_.isEmpty(params)){
+        for(var i in params){
+            queryStringParams.push(encodeURIComponent(i)+'='+encodeURIComponent(params[i]));
```
Comment (markwallsgrove) [inline index.js]: need a check here: `if (params.hasOwnProperty(i)) { }`

Comment (rgubby) [inline index.js]: doesn't the `for(var i i params)` give a list of properties back - not sure what the hasOwnProperty will give me?

Comment (markwallsgrove) [inline index.js]: http://jshint.com/docs/options/#forin

Comment (rgubby) [inline index.js]: :+1: ta

File: index.js [hunk 2]
```diff
@@ -52,12 +106,29 @@ BabelClient.prototype.getTargetFeed = function(target, token, hydrate, callback)
...
 
     hydrate = hydrate || false;
 
+    // Build up a query string
+    var queryString = '',
+       queryStringParams = [];
+
+    if(!_.isEmpty(params)){
+        for(var i in params){
+            queryStringParams.push(encodeURIComponent(i)+'='+encodeURIComponent(params[i]));
```
Comment (markwallsgrove) [inline index.js]: ditto with `hasOwnProperty`

File: README.md [hunk 3]
```diff
@@ -25,11 +25,26 @@ To use any of the Babel client functions, you must have a Persona token. Read th
...
+Get a feed based on a target. Now supports passing through of `offset`, `limit` and `delta_token` in params.
+```javascript
+var target = 'stay-on-target';
+var token = req.personaClient.getToken(req);
+var hydrate = true;
+var params = {
+    delta_token:'123'
+};
+babelClient.getTargetFeed(target, token, hydrate, params, function(error, results){
     // do stuff
```
Comment (pads) [inline README.md]: It might be nice to have a simple example of what to do with the results e.g. what format they come back in?

File: index.js [hunk 4]
```diff
@@ -35,15 +35,67 @@ var BabelClient = function (config) {
...
+        }
+    };
+
+    this.debug(JSON.stringify(requestOptions));
+
+    request(requestOptions, function(err, response){
+        if(err){
+            callback(err);
+        } else if(response.statusCode && response.statusCode !== 204)
+        {
```
Comment (pads) [inline index.js]: Dropped bracket looks a bit PHPish

File: index.js [hunk 5]
```diff
@@ -35,15 +35,67 @@ var BabelClient = function (config) {
...
+        callback = params;
+        params = {};
+    }
+
+    // Build up a query string
+    var queryString = this._queryStringParams(params);
+
+    var requestOptions = {
+        method: "HEAD",
+        url: this.config.babel_host+':'+this.config.babel_port+'/feeds/targets/'+md5(target)+'/activity/annotations' + (!_.isEmpty(queryString) ? '?'+queryString : ''),
```
Comment (markwallsgrove) [inline index.js]: would be good to break this over multiple lines

File: index.js [hunk 6]
```diff
@@ -35,15 +35,67 @@ var BabelClient = function (config) {
 };
 
 /**
+ * Make a HEAD request to check for new annotations
+ * @param {string} target
+ * @param {string} token
+ * @param {object} params
+ * @param {string} params.delta_token Using a delta_token tells you if there have been new annotations since another one
+ * @callback callback
```
Comment (markwallsgrove) [inline index.js]: missing `@throws`

File: index.js [hunk 7]
```diff
@@ -35,15 +35,67 @@ var BabelClient = function (config) {
...
+
+/**
  * Get a feed based off a target identifier. Return either a list of feed identifiers, or hydrate it and
  * pass back the data as well
  *
  * @param {string} target Feed target identifier
  * @param {string} token Persona token
  * @param {boolean} hydrate Gets a fully hydrated feed, i.e. actually contains the posts
+ * @param {object} params Query params to pass into the call to Babel
  * @callback callback
```
Comment (markwallsgrove) [inline index.js]: missing `@throws`

File: index.js [hunk 8]
```diff
@@ -52,12 +104,21 @@ BabelClient.prototype.getTargetFeed = function(target, token, hydrate, callback)
...
     var self = this;
 
     hydrate = hydrate || false;
 
+    // Build up a query string
+    var queryString = this._queryStringParams(params);
+
     var requestOptions = {
-        url: this.config.babel_host+':'+this.config.babel_port+'/feeds/targets/'+md5(target)+'/activity/annotations'+((hydrate === true) ? '/hydrate' : ''),
+        url: this.config.babel_host+':'+this.config.babel_port+'/feeds/targets/'+md5(target)+'/activity/annotations'+((hydrate === true) ? '/hydrate' : '') + (!_.isEmpty(queryString) ? '?'+queryString : ''),
```
Comment (markwallsgrove) [inline index.js]: can this be broken over multiple lines?

**Comment by pads** [comment]
My babel knowledge is still lacking but looks good :+1:

File: index.js (hunk 4 above)
Comment (rgubby) [inline index.js]: Yikes!

File: index.js (hunk 6 above)
Comment (rgubby) [inline index.js]: done

File: index.js (hunk 7 above)
Comment (rgubby) [inline index.js]: and done

**Pull request merged**